crates/solvers/src/domain/solver/dex/mod.rs

```diff
@@ -86,40 +101,61 @@ impl Dex {
+    ) -> Result<dex::Swap, infra::dex::Error> {
+        let slippage = self.slippage.relative(&dex_order.amount(), tokens);
+        self.dex
+            .swap(dex_order, &slippage, tokens, gas_price)
+            .await
+            .map_err(|err| {
+                match &err {
+                    err @ infra::dex::Error::NotFound => {
+                        if order.partially_fillable {
+                            // Only adjust the amount to try next if we are sure the API worked
+                            // correctly yet still wasn't able to provide a
+                            // swap.
+                            self.fills.reduce_next_try(order.uid);
+                        } else {
+                            tracing::debug!(?err, "skipping order");
+                        }
+                    }
+                    err @ infra::dex::Error::OrderNotSupported => {
+                        tracing::debug!(?err, "skipping order")
+                    }
+                    err @ infra::dex::Error::RateLimited => {
+                        tracing::debug!(?err, "encountered rate limit")
+                    }
+                    infra::dex::Error::Other(err) => tracing::warn!(?err, "failed to get swap"),
+                }
+                err
+            })
+    }
+
     async fn solve_order(
         &self,
         order: order::UserOrder<'_>,
         tokens: &auction::Tokens,
         gas_price: auction::GasPrice,
     ) -> Option<solution::Solution> {
         let order = order.get();
-        let swap = {
-            let order = self.fills.dex_order(order, tokens)?;
-            let slippage = self.slippage.relative(&order.amount(), tokens);
-            self.dex.swap(&order, &slippage, tokens, gas_price).await
-        };
-
-        let swap = match swap {
-            Ok(swap) => swap,
-            Err(err @ infra::dex::Error::NotFound) => {
-                if order.partially_fillable {
-                    // Only adjust the amount to try next if we are sure the API worked correctly
-                    // yet still wasn't able to provide a swap.
-                    self.fills.reduce_next_try(order.uid);
-                } else {
-                    tracing::debug!(?err, "skipping order");
-                }
-                return None;
-            }
-            Err(err @ infra::dex::Error::OrderNotSupported) => {
-                tracing::debug!(?err, "skipping order");
-                return None;
-            }
-            Err(infra::dex::Error::Other(err)) => {
-                tracing::warn!(?err, "failed to get swap");
-                return None;
-            }
-        };
+        let dex_order = self.fills.dex_order(order, tokens)?;
+        let swap = self
+            .rate_limiter
+            .execute_with_back_off(
+                self.try_solve(order, &dex_order, tokens, gas_price),
+                |result| matches!(result, Err(infra::dex::Error::RateLimited)),
```

Comment:
Maybe move this part into try_solver so that we have one method which returns the high level Result that we care about.

Reply:
Moved